functions/database/database.js

```diff
@@ -0,0 +1,316 @@
+ * that have the following structure:
+ * {
+ *   releaseName: string,
+ *   releaseOperator: string,
+ *   codeFreezeDate: Date,
+ *   releaseDate: Date,
+ * }
+ *
+ * @param {Object} newReleases - Releases to store in Firestore
+ */
+async function addReleases(newReleases) {
+  // Validate the structure of the new releases before we store
+  validateNewReleaseStructure(newReleases);
+
+  const batch = db.batch();
+
+  newReleases.forEach((release) => {
+    const newReleaseRef = db.collection("releases").doc();
+    const snapshotBranchName = `releases/${release.releaseName}`;
+    const releaseBranchName = `releases/${release.releaseName}.release`;
+    const snapshotBranchLink = `${REPO_URL}/tree/${snapshotBranchName}`;
+    const releaseBranchLink = `${REPO_URL}/tree/${releaseBranchName}`;
+    const releaseData = {
+      state: RELEASE_STATES.SCHEDULED,
+      releaseName: release.releaseName,
+      releaseOperator: release.releaseOperator,
+      codeFreezeDate: release.codeFreezeDate,
+      releaseDate: release.releaseDate,
+      snapshotBranchName: snapshotBranchName,
+      snapshotBranchLink: snapshotBranchLink,
+      releaseBranchName: releaseBranchName,
+      releaseBranchLink: releaseBranchLink,
+      isComplete: false,
+      buildArtifactStatus: "",
+      buildArtifactConclusion: "",
+      buildArtifactLink: "",
+      buildArtifactJobId: "",
+    };
+    batch.set(newReleaseRef, releaseData);
+  });
+
+  await batch.commit();
+}
+
+/**
+   * Delete the releases in Firestore with code freeze dates that are in the
+   * future. This is intented to be used when scheduling new upcoming releases,
+   * but we need to delete the old upcoming releases before.
+   */
+async function deleteUpcomingReleases() {
+  const today = Timestamp.now();
+  const releasesRef = db.collection("releases");
+  const query = releasesRef.where("codeFreezeDate", ">", today);
+  const querySnapshot = await query.get();
+  const batch = db.batch();
+
+  querySnapshot.forEach((doc) => {
+    batch.delete(doc.ref);
+  });
+
+  await batch.commit();
+}
+
+/**
+   * Update a release in Firestore.
+   *
+   * @param {string} releaseId The ID of the release to update.
+   * @param {string} releaseData The new data to set for the release.
+   * @throws {Error} If the release state update fails.
+   */
+async function updateRelease(releaseId, releaseData) {
+  validateRelease(releaseData);
+  const releaseDoc = db.collection("releases").doc(releaseId);
+  await releaseDoc.update(releaseData);
+}
+
+/**
+ * Updates the state of a release in the database.
+ *
+ * @param {string} releaseId The ID of the release to update
+ * @param {string} newState The new state to set for the release
+ * @throws {Error} If the release state update fails, or if the
+ * release state is invalid.
+ */
+async function updateReleaseState(releaseId, newState) {
+  // Verify that the state is one of the valid states
```

Comment:
No need for the comment, the error message provides enough information

Reply:
Resolved in https://github.com/firebase/firebase-release-dashboard/pull/8/commits/a32e049b16404c7818f488231642a11b343b3068